lightning/src/ln/channelmanager.rs

```diff
@@ -3477,6 +3477,35 @@ where
 		}
 	}
 
+	/// This is a variant of [`ChannelManager::fail_htlc_backwards`] that allows you to specify the
+	/// reason for the failure.
+	/// 
+	/// Indicates that the preimage for payment_hash is unknown or the received amount is incorrect
+	/// after a PaymentClaimable event, failing the HTLC back to its origin and freeing resources
+	/// along the path (including in our own channel on which we received it).
+	///
+	/// Note that in some cases around unclean shutdown, it is possible the payment may have
+	/// already been claimed by you via [`ChannelManager::claim_funds`] prior to you seeing (a
+	/// second copy of) the [`events::Event::PaymentClaimable`] event. Alternatively, the payment
+	/// may have already been failed automatically by LDK if it was nearing its expiration time.
+	///
+	/// While LDK will never claim a payment automatically on your behalf (i.e. without you calling
+	/// [`ChannelManager::claim_funds`]), you should still monitor for
+	/// [`events::Event::PaymentClaimed`] events even for payments you intend to fail, especially on
+	/// startup during which time claims that were in-progress at shutdown may be replayed.
+	pub fn fail_htlc_backwards_with_reason(&self, payment_hash: &PaymentHash, reason: &HTLCFailReason) {
```

Comment:
I'm confused why CI is passing on this PR, because `HTLCFailReason` is in a private module 🤔 

In any case, I think we should change this to be parameterized by a `failure_code: u16` and `failure_data: Vec<u8>` and revert the `onion_utils` changes

Reply:
Ah, makes sense!